File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int data;
 *     TreeNode *left;
 *     TreeNode *right;
 *      TreeNode(int val) : data(val) , left(nullptr) , right(nullptr) {}
 * };
 **/

class Solution {
private:
    void inorder(TreeNode* root, int &smallCounter, int k, int &smallAns){
        if(!root || smallAns) return;

        inorder(root->left, smallCounter, k , smallAns);
        smallCounter++;
        if(smallCounter == k){
            smallAns = root->val;
            return;
        }
        inorder(root->right, smallCounter, k , smallAns);
    }
public:
	int kthSmallest(TreeNode* root,int k){
        int smallCounter = 0;
        int smallAns = 0;
        inorder(root, smallCounter, k, smallAns);
        return smallAns;
	}
};

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.py
class Solution:
    def kthSmallest(self, root: TreeNode | None, k: int) -> int:
        cnt = 0

        def dfs(root):
            nonlocal cnt

            if not root:
                return None

            # Left
            result = dfs(root.left)
            if result is not None:
                return result

            # Current
            cnt += 1
            if cnt == k:
                return root.val

            # Right
            return dfs(root.right)

        return dfs(root)